### PR TITLE
Fix race that could lead to pg wire messages being out-of-order

### PR DESCRIPTION
**Pull request by mfussenegger**
## Summary of the changes / Why this improves CrateDB

The `completionFuture` of a `ResultReceiver` was used to both:

- Write out pending messages
- Trigger the next execution if there are multiple deferred statements

There was a race where the next execution could write out
`commandComplete` _before_ the "write pending messages" routine
triggered, leading to out-of-order messages on the client side.

This caused the JDBC client to fail with:

	java.util.NoSuchElementException
		at __randomizedtesting.SeedInfo.seed([FAA2E13C629A53DB:BF7BE6AFD08EA59]:0)
		at java.base/java.util.ArrayDeque.removeFirst(ArrayDeque.java:362)
		at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2129)
		at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:356)
		at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:496)
		at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:413)
		at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:333)
		at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:319)
		at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:295)
		at org.postgresql.jdbc.PgStatement.executeUpdate(PgStatement.java:268)

To remove the race this moves the "write pending messages" trigger into
the ResultReciver to ensure it's run _before_ triggering the completion
future.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by mfussenegger** [comment]
Looks like this breaks npgsql, need to have another look

**Comment by mfussenegger** [comment]
> Looks like this breaks npgsql, need to have another look

Pushed a fixup to solve it. The problem was that `writePendingMessages` wrote out the full hierarchy (traversing over the parents) instead of only the messages related to a given `execute` cycle.

File: server/src/main/java/io/crate/protocols/postgres/Messages.java [hunk 1]
```diff
@@ -107,7 +107,7 @@ static ChannelFuture sendCommandComplete(Channel channel, String query, long row
         buffer.writeByte('C');
         buffer.writeInt(length);
         writeCString(buffer, commandTagBytes);
-        ChannelFuture channelFuture = channel.writeAndFlush(buffer);
+        ChannelFuture channelFuture = channel.write(buffer);
         if (LOGGER.isTraceEnabled()) {
             channelFuture.addListener((ChannelFutureListener) future -> LOGGER.trace("sentCommandComplete"));
         }
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/protocols/postgres/Messages.java:113]: minor: can we have a misleading log message saying "sent" before the message got flushed?

**Review by BaurzhanSakhariev — APPROVED** [review]
👍 added a minor comment

File: server/src/main/java/io/crate/protocols/postgres/Messages.java (hunk 1 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/protocols/postgres/Messages.java:113]: The futures only  trigger once the message is flushed out, so this is save. (We do the same already for most other messages)

**Comment by mfussenegger** [comment]
retest this please

flaky `test_subscription_state_order`

**Pull request merged**